### PR TITLE
Track connection state, reconnect on error, outgoing message queues

### PR DESCRIPTION
**Pull request by callebtc**
This PR adds tracking for the conenction state of a Relay. 

In particular, when using multiple ralays of which some may be faulty and/or just slow, the method `Relay.publish` would try to publish independent of the connection state and raise `WebSocketConnectionClosedException: Connection is already closed.` 

This means that all publishing attempts on other relays after the faulty one would obviously also fail.

#### New or changed attributes

In this PR, I introduce a series of new states for the `Relay` class: 

- `Relay.connected` tracks whether we are connected to the Relay.
- `Relay.reconnect` determines whether we should attempt to reconnect on a dropped connection.
- `Relay.error_counter` counts the number of errors received from a relay.
- `Relay.error_threshold` decides how often we want to attempt to reconnect on an error.
- `Relay.ssl_options` stores the SSL options that we need for a reconnect.

#### New or changed methods

- `Relay.check_reconnect` is called to check whether conditions for reconnect are met
- `Relay._on_error` increments the `error_counter` and calls `check_reconnect` if it is below the `error_threshold`

### Message `Queue`
- `Relay` has a `queue` now which will publish messages only if its state is `connected`.
- The queue worker is called in the `RelayManager` in a separate thread.

### Notes

The placing of these methods might be a bit too low-level for your taste. I'm open for suggestions. It could also be moved to `RelayManager` which would mean that 

- We would need to be a lot more callbacks from `Relay` to the `RelayManager` (for errors, connects, disconnects, ...)
- We would introduce these lower-level things there.

It feels to me that this is better suited for `Relay` since it really only matters for one specific relay.

If you think this is out of place, feel free to close the PR. I'm using this now in a [fork](https://github.com/callebtc/python-nostr) for the [Cashu](https://github.com/cashubtc/cashu) client.

### CODE REVIEW

**Comment by callebtc** [comment]
@jeffthibault can you have a look at this?

File: nostr/relay.py [hunk 1]
```diff
@@ -48,8 +56,19 @@ def connect(self, ssl_options: dict=None, proxy: dict=None):
...
 
+    def check_reconnect(self):
+        try:
+            self.close()
+        except:
+            pass
+        self.connected = False
+        if self.reconnect:
+            time.sleep(1)
+            self.connect(self.ssl_options)
```
Comment (jeffthibault) [inline nostr/relay.py:67]: This method creates an infinite loop as long as the connection is alive. Should this be called in a thread?

Comment (callebtc) [reply nostr/relay.py:67]: Does it? The method is only called `_on_error()` which means the connection dropped already. 

If you look into the `websocket`, it calls `websocket._app.teardown()` on error which calls `thread.join()` basically getting rid of the thread.

**Comment by callebtc** [comment]
I've updated this PR to your newest `main` to also reuse the `proxy` setting upon a reconnect.

File: nostr/relay.py (hunk 1 above)
Comment (jeffthibault) [reply nostr/relay.py]: `self.connect` will call `ws.run_forever`, which is where the infinite loop is created. It seems like that would block the main thread.

Comment (callebtc) [reply nostr/relay.py]: > `self.connect` will call `ws.run_forever`, which is where the infinite loop is created. It seems like that would block the main thread.

Yes, but it actually doesn't run forever. It tears down the thread it's running in `_on_error`. At least that's what I gather from the code. 

I will recheck this to be sure and track the number of threads. There are lots of relays with errors these days. Will report back in a few minutes.

Comment (jeffthibault) [reply nostr/relay.py]: Cool, thanks. No doubt there are lots of relays with errors now so we absolutely need this.

Comment (callebtc) [reply nostr/relay.py]: Checked. `ws.run_forever` does not run forever. It tears down the Thread on error.

**Comment by callebtc** [comment]
HI @jeffthibault, I've kept working on this PR in my own fork and have added message queues to the relays.

See here: https://github.com/callebtc/python-nostr/pull/6/commits/1812b0128f230a7b8be5df48ff7077edc6b64484, `relay.py` and `relay_manager.py`.

That means no sleeps and client-side connection state checks anymore. A client can just connect and start publishing messages. The `relay_manager` calls a separate Thread in `Relay` that works off the queue full of messages.

Let me know if you'd be interested in that, I could add that to this PR or in a separate one. I think it fits nicely in this PR though.

**Comment by jeffthibault** [comment]
@callebtc I love the queues in each relay. I was planning to add something similar myself as a follow up but feel free to incorporate it here. It does fit nicely.

File: nostr/relay.py (hunk 1 above)
Comment (jeffthibault) [reply nostr/relay.py]: Thanks for checking

**Comment by callebtc** [comment]
@jeffthibault, mad conflicts after your refactor...

**Comment by jeffthibault** [comment]
@callebtc my apologies. Will you be able to resolve?

**Comment by jeffthibault** [comment]
@callebtc I resolved the merge conflicts. Can you verify things are working on your end?

**Comment by callebtc** [comment]
LGTM! Thank you for resolving the conflicts.

I've added a hotfix for the initialization value of `PublicKey()` otherwise it can't be intialized with `PublicKey().from_npub(...)`

**Comment by callebtc** [comment]
Sorry for the formatting, removed it in the last commit (I'm using black).

**Comment by jeffthibault** [comment]
> I've added a hotfix for the initialization value of `PublicKey()` otherwise it can't be intialized with `PublicKey().from_npub(...)`

I don't think this is necessary.  

`from_npub` is a classmethod so you should use it like so:
```pubkey = PublicKey.from_npub(...)```

File: nostr/relay.py [hunk 2]
```diff
@@ -60,8 +65,27 @@ def connect(self):
     def close(self):
         self.ws.close()
 
+    def check_reconnect(self):
+        try:
+            self.close()
+        except:
+            pass
+        self.connected = False
+        if self.reconnect:
```
Comment (jeffthibault) [inline nostr/relay.py:74]: Where does `self.reconnect` change? I only see it getting set to `True` in `__post_init__`.

Is it actually needed?

Comment (callebtc) [reply nostr/relay.py:74]: It was meant as a configuration in the class init so the "user" can choose whether they want reconnect attempts or not.

**Comment by callebtc** [comment]
> > I've added a hotfix for the initialization value of `PublicKey()` otherwise it can't be intialized with `PublicKey().from_npub(...)`
> 
> I don't think this is necessary.  
> 
> `from_npub` is a classmethod so you should use it like so:
> ```pubkey = PublicKey.from_npub(...)```

Oh thank you for the note, then I was using it wrong. However, now it's the same as `PrivateKey`.

**Comment by jeffthibault** [comment]
> However, now it's the same as `PrivateKey`.

Not sure what this means. Can you give some more details?

**Comment by callebtc** [comment]
> > However, now it's the same as `PrivateKey`.
> 
> Not sure what this means. Can you give some more details?

Sure. Here is the definition of PrivateKey:

```python
class PrivateKey:
    def __init__(self, raw_secret: bytes = None) -> None:
```

Here is `PublicKey` (before)

```python
class PublicKey:
    def __init__(self, raw_bytes: bytes) -> None:
```

I've added `raw_bytes: bytes = None` (after).

**Comment by callebtc** [comment]
Resolved conflicts from latest PR.

**Comment by jeremywhelchel** [comment]
Is there a reason to not use the reconnection behavior already provided by WebSocketApp? By passing a dispatcher and connection interval to run_forever() as shown here: https://github.com/websocket-client/websocket-client#long-lived-connection

In my experience I have seen both the intentional disconnects (eg "cloud flare restarting web socket proxy") and lost connections. The websocket library only reconnects the latter, but it might be good to handle both these cases somewhat differently.

**Comment by jeffthibault** [comment]
@jeremywhelchel it would require adding another dependency (`rel`) which I prefer to avoid when possible. I'm not totally against it. Happy to experiment with it and see if it's better.

**Comment by jeffthibault** [comment]
LGTM

**Pull request merged**